---
title: "Practical 2 - Population Structure Inference"
author: ""
date: ""
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---
Before you begin:

* Make sure that R is installed on your computer
* For this lab, we will use the following R libraries:
```{r load-libs, message=FALSE, warning=FALSE}
library(data.table)
library(dplyr)
library(bigsnpr)
library(ggplot2)
```

The R template to do the exercises is [here](https://github.com/joellembatchou/SISG2024_Association_Mapping/tree/master/code).

## Introduction
We will be working with a subset of the genotype data from the Human Genome Diversity Panel (HGDP) and HapMap.

The file “[YRI_CEU_ASW_MEX_NAM.bed](https://github.com/joellembatchou/SISG2024_Association_Mapping/tree/master/data)” is a binary file in PLINK BED format with accompanying BIM and FAM files.
It contains the genotype data at autosomal SNPs (i.e. chromosomes 1-22) for:

* Native American samples from HGDP
* Four population samples from HapMap: 
  + Yoruba in Ibadan, Nigeria (YRI)
  + Utah residents with ancestry from
Northern and Western Europe (CEU)
  + Mexican Americans in Los Angeles, California (MXL)
  + African Americans from the south-western United States (ASW)

**File with ancestry labels assignment for each sample**: [Population_Sample_Info.txt](https://raw.githubusercontent.com/joellembatchou/SISG2024_Association_Mapping/master/data/Population_Sample_Info.txt) 

### Data preparation
Let’s first load the HGDP data into the R session. We need to define the path to the directory containing the PLINK BED and the ancestry label files (change the path to the file location).
```{r echo=FALSE}
# change this to the directory on your machine
HGDP_dir <- "/Users/joelle.mbatchou/SISG/2024/SISG2024_Association_Mapping/data/" 
```

```{r eval=FALSE}
# change this to the directory on your machine
HGDP_dir <- "/SISGM19/data/" 
```

Also specify the path to the PLINK2 binary
```{r eval=FALSE}
plink2_binary <- "/SISGM19/bin/plink2" 
```

We can now read the PLINK BED and FAM files (recall the BED file is a binary file):
```{r}
HGDP_bim <- fread(sprintf("%s/YRI_CEU_ASW_MEX_NAM.bim", HGDP_dir), header = FALSE)
head(HGDP_bim, 3)
HGDP_fam <- fread(sprintf("%s/YRI_CEU_ASW_MEX_NAM.fam", HGDP_dir), header = FALSE)
head(HGDP_fam, 3)
```

When reading the ancestry label file, we need to make sure the order of samples matches that in the PLINK data:
```{r}
HGDP_ancestry_df <- fread(sprintf("%s/Population_Sample_Info.txt", HGDP_dir))
HGDP_ancestry_df <- left_join(HGDP_fam[,c("V1","V2")], HGDP_ancestry_df, by = c("V1" = "FID", "V2" = "IID"))
head(HGDP_ancestry_df, 3)
```

## Exercises
Here are some things to look at:

1. Examine the dataset:
    * How many samples are present? `str(HGDP_fam)`
    * How many SNPs? `str(HGDP_bim)`
    * What is the number of samples in each population? `table(HGDP_ancestry_df$Population)`

2. Get the first 10 principal components (PCs) in PLINK using all SNPs. 
```{r eval=FALSE}
cmd <- sprintf("%s --bfile %s/YRI_CEU_ASW_MEX_NAM --pca 10 --out pca_plink", plink2_binary, HGDP_dir)
system(cmd)
```

This generates two files `pca_plink.eigenvec` containing the PCs (eigenvectors), and `pca_plink.eigenval` containing the top eigenvalues.

3. Read in the PCs in R and make a scatterplot of the first two PCs with each point colored by population membership.
```{r eval=FALSE}
PC_df <- left_join(fam_pop_info, fread("pca_plink.eigenvec"), by = c("V1" = "#FID", "V2" = "IID"))
with(PC_df, plot(x=PC1, y=PC2, color = Population))
```

4. Interpret the first two PCs, what ancestries are they reflecting?

5. Read in the eigenvalues and a scree plot corresponding to the first 10 PCs. Estimate the proportion of variance explained by the first two PCs.
```{r eval=FALSE}
eigenvalues_df <- fread("pca_plink.eigenval", header = FALSE)
with(eigenvalues_df, plot(x=1:10, y=V1, xlab = "PC", ylab = "Eigenvalue"))
```

6. Now redo Question 2 above using the [`bigsnpr` R package](https://privefl.github.io/bigsnpr/reference/index.html) specifying a $r^2$ threshold of 0.2 (i.e. LD pruning) as well as a minimum minor allele count (MAC) of 20. The basic command would look like
```{r eval = FALSE}
obj.bed <- bed(bedfile = sprintf("%s/YRI_CEU_ASW_MEX_NAM.bed", HGDP_dir))
pca.bigsnpr <- bed_autoSVD(
  obj.bed, 
  thr.r2 = 0.1, # R^2 threshold
  k = 10, # number of PCs
  min.mac = 10 # minimum minor allele count (MAC) filter
)
```

* You can evaluate the PCA results using
```{r, eval = FALSE}
# plot PC2 vs PC1
plot(pca.bigsnpr, type = "scores", scores = 1:2)
# scree plot
plot(pca.bigsnpr) 
# plot SNP loadings for the first 5 PCs
plot(pca.bigsnpr, type = "loadings", scores = 1:5, coeff = 0.4)
```

* Make a scatter plot of the first two principal components (PCs) with each point colored according to population membership. Does the plot change from the one in Question 2?
```{r, eval = FALSE}
plot(pca.bigsnpr, type = "scores", scores = 1:2) +
  aes(color = fam_pop_info$Population) +
  labs(color = "Population")
```

* Check the SNP loadings for the first 10 PCs.
  
7. Predict the proportional Native American and European Ancestry for the HapMap
MXL from the PCA output in Question 6 *using one of the principal components*. (Which PC is most appropriate for this analysis?) Assume that the HapMap MXL have negligible African Ancestry.

8. Make a barplot of the proportional ancestry estimates from question 4.

### *Extra*
9. Check if there are samples related 2nd degree or closer. If so, run PCA as in Question 6 removing these samples then project the remaining samples onto the PC space. 

* The command to check relatedness is:
```{r, eval = FALSE}
# check for 3rd degree relateds or closer
relatedness_info <- snp_plinkKINGQC(
  plink2.path = plink2_binary, 
  bedfile.in = sprintf("%s/YRI_CEU_ASW_MEX_NAM.bed", HGDP_dir), 
  thr.king = 2^-3.5, # threshold for 3rd degree relateds
  make.bed = FALSE
)
```
This returns a data frame which contains all pairs of individuals related 3rd degree or closer. 

* We can then remove them when calling `bed_autoSVD()` using the `ind.row` argument.
```{r, eval = FALSE}
ind.rel <- match(c(relatedness_info$IID1, relatedness_info$IID2), obj.bed$fam$sample.ID)
ind.norel <- rows_along(obj.bed)[-ind.rel]
obj.svd2 <- bed_autoSVD(
  obj.bed, 
  thr.r2 = 0.1, # R^2 threshold
  k = 10, # number of PCs
  min.mac = 10, # minimum minor allele count (MAC) filter
  ind.row = ind.norel
)
```

* Use `bed_projectSelfPCA()` to project related samples on the PC space.
*(Hint: This [tutorial document](https://privefl.github.io/bigsnpr/articles/bedpca.html) from `bigsnpr` will be helpful -- see the last section 'Project remaining individuals')*

